common/src/main/java/org/apache/cassandra/sidecar/common/JmxClient.java

```diff
@@ -0,0 +1,241 @@
+
+import org.jetbrains.annotations.VisibleForTesting;
+
+/**
+ * A simple wrapper around a JMX connection that makes it easier to get proxy instances.
+ */
+public class JmxClient implements NotificationListener
+{
+    public static final String JMX_SERVICE_URL_FMT = "service:jmx:rmi://%s/jndi/rmi://%s:%d/jmxrmi";
+    public static final String REGISTRY_CONTEXT_SOCKET_FACTORY = "com.sun.jndi.rmi.factory.socket";
+    private final JMXServiceURL jmxServiceURL;
+    private MBeanServerConnection mBeanServerConnection;
+    private final Map<String, Object> jmxEnv;
+    private boolean connected = false;
+
+    /**
+     * Creates a new client with the provided {@code host} and {@code port}.
+     *
+     * @param host the host of the JMX service
+     * @param port the port of the JMX service
+     */
+    public JmxClient(String host, int port)
+    {
+        this(host, port, null, null, false);
+    }
+
+    /**
+     * Creates a new client with the provided parameters
+     *
+     * @param host      the host of the JMX service
+     * @param port      the port of the JMX service
+     * @param role      the JMX role used for authentication
+     * @param password  the JMX role password used for authentication
+     * @param enableSSl true if SSL is enabled for JMX, false otherwise
+     */
+    public JmxClient(String host, int port, String role, String password, boolean enableSSl)
+    {
+        this(buildJmxServiceURL(host, port), role, password, enableSSl);
+    }
+
+    @VisibleForTesting
+    JmxClient(JMXServiceURL jmxServiceURL)
+    {
+        this(jmxServiceURL, null, null, false);
+    }
+
+    @VisibleForTesting
+    JmxClient(JMXServiceURL jmxServiceURL, String role, String password)
+    {
+        this(jmxServiceURL, role, password, false);
+    }
+
+    private JmxClient(JMXServiceURL jmxServiceURL, String role, String password, boolean enableSsl)
+    {
+        this.jmxServiceURL = jmxServiceURL;
+
+        jmxEnv = new HashMap<>();
+        if (role != null && password != null)
+        {
+            String[] credentials = new String[]{ role, password };
+            jmxEnv.put(JMXConnector.CREDENTIALS, credentials);
+        }
+        jmxEnv.put(REGISTRY_CONTEXT_SOCKET_FACTORY, getRMIClientSocketFactory(enableSsl));
+    }
+
+    /**
+     * Applies a function through the JMX bean proxy
+     *
+     * @param clientClass the management interface that the MBean exports, which will
+     *                    also be implemented by the returned proxy
+     * @param remoteName  the name of the MBean within {@code connection} to forward to
+     * @param func        the remote function that will be executed through the proxy
+     * @param <TClient>   the type of the proxy client
+     * @param <TResult>   the type of the result returned by the proxy
+     * @return the result of the remote function after being executed via the proxy
+     */
+    public <TClient, TResult> TResult apply(Class<TClient> clientClass,
```

Comment:
Generic types named in the form that looks like class names are misleading and confusing. Can you just do `C` for client and `R` for result?